File: bench/algorithm/binarytrees/3.cs
namespace BenchmarkGameBTrees
{
    /*
          The Computer Language Benchmarks Game
          https://salsa.debian.org/benchmarksgame-team/benchmarksgame/ 

          contributed by Marek Safar  
          optimized by kasthack
          *reset*
    */
    using System;
    using System.Threading;
    using System.Threading.Tasks;

    class BinaryTrees
    {
        const int minDepth = 4;
        public static void Main(String[] args)
        {
            int n = 0;
            if (args.Length > 0) n = Int32.Parse(args[0]);
            int maxDepth = Math.Max(minDepth + 2, n);
            int stretchDepth = maxDepth + 1;
            int check = (TreeNode.bottomUpTree(stretchDepth)).itemCheck();
            Console.WriteLine("stretch tree of depth {0}\t check: {1}", stretchDepth, check);
            TreeNode longLivedTree = TreeNode.bottomUpTree(maxDepth);
            for (int depth = minDepth; depth <= maxDepth; depth += 2)
            {
                int iterations = 1 << (maxDepth - depth + minDepth);
                check = 0;

                Parallel.For(1, iterations + 1,
                    () => 0,
                    (i, loop, localCheck) =>
                    {
                        return localCheck + 
                            (TreeNode.bottomUpTree(depth)).itemCheck();
                    },
                    localCheck => Interlocked.Add(ref check, localCheck)
                );

                Console.WriteLine("{0}\t trees of depth {1}\t check: {2}",
                    iterations, depth, check);
            }
            Console.WriteLine("long lived tree of depth {0}\t check: {1}",
                maxDepth, longLivedTree.itemCheck());
        }

        class TreeNode
        {
            private TreeNode left, right;

            internal static TreeNode bottomUpTree(int depth)
            {
                TreeNode t;
                ChildTreeNodes(out t, depth);
                return t;
            }
            static void ChildTreeNodes(out TreeNode node, int depth)
            {
                node = new TreeNode();
                if (depth > 0)
                {
                    ChildTreeNodes(out node.left, depth - 1);
                    ChildTreeNodes(out node.right, depth - 1);
                }
            }
            internal int itemCheck()
            {
                if (right == null) return 1;
                else return 1 + left.itemCheck() + right.itemCheck();
            }
        }
    }
}

File: bench/algorithm/binarytrees/3.rs
use std::rc::Rc;

struct TreeNode {
    l: Option<Rc<TreeNode>>,
    r: Option<Rc<TreeNode>>,
}

impl TreeNode {
    fn check(&self) -> i32 {
        let mut ret = 1;
        if let Some(l) = &self.l {
            ret += l.check();
        }
        if let Some(r) = &self.r {
            ret += r.check();
        }
        ret
    }

    fn create<'r>(depth: i32) -> Rc<TreeNode> {
        if depth > 0 {
            Rc::new(TreeNode {
                l: Some(Self::create(depth - 1)),
                r: Some(Self::create(depth - 1)),
            })
        } else {
            Rc::new(TreeNode { l: None, r: None })
        }
    }
}

const MIN_DEPTH: i32 = 4;

fn main() {
    let n = std::env::args_os()
        .nth(1)
        .and_then(|s| s.into_string().ok())
        .and_then(|n| n.parse().ok())
        .unwrap_or(10);

    let max_depth = if MIN_DEPTH + 2 > n { MIN_DEPTH + 2 } else { n };
    {
        let depth = max_depth + 1;
        let tree = TreeNode::create(max_depth + 1);

        println!("stretch tree of depth {}\t check: {}", depth, tree.check());
    }

    let long_lived_tree = TreeNode::create(max_depth);

    for d in (MIN_DEPTH..max_depth + 1).step_by(2) {
        let iterations = 1 << ((max_depth - d + MIN_DEPTH) as u32);
        let mut chk = 0;
        for _i in 0..iterations {
            let a = TreeNode::create(d);
            chk += a.check();
        }
        println!("{}\t trees of depth {}\t check: {}", iterations, d, chk)
    }

    println!(
        "long lived tree of depth {}\t check: {}",
        max_depth,
        long_lived_tree.check()
    );
}
